src/vs/editor/common/modes.ts

```diff
@@ -392,7 +392,7 @@ export interface ReferenceProvider {
 	/**
 	 * Provide a set of project-wide references for the given position and document.
 	 */
-	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, token: CancellationToken): Location[] | Thenable<Location[]>;
+	provideReferences(model: editorCommon.IReadOnlyModel, position: Position, context: ReferenceContext, progress: (value: Location) => void, token: CancellationToken): Location[] | Thenable<Location[]>;
```

Comment:
> stick to JSON Patches. The caller would have to apply them

JSON patch seems like a wire format implementation detail. I am not convinced that we want that to leak into the vscode codebase.

> always call progress with the not-finished result (what the promise will eventually resolve to)

Possible, but that does result in a lot of overhead for large result sets. It is local so not as bad, but still seems less than ideal.

What I was thinking was a progress callback like this
```
(value: Location | PartialLocation) => void
```

Then the extension just calls like this
```
// Notify progress of partial results
progress(new PartialLocation("1"));
progress(new PartialLocation("2"));
progress(new PartialLocation("3"));

// Later notify progress of full results
progress(new Location("1", "some other data 1"));
progress(new Location("2", "some other data 2"));
progress(new Location("3", "some other data 3"));
```

As long as each location is uniquely identifiable (they are by dir + file + range), then vscode can be smart enough to have complete locations overwrite the associated PartialLocation if it exists.

Thoughts?